File: 2692-take-gifts-from-the-richest-pile/2692-take-gifts-from-the-richest-pile.rb
# @param {Integer[]} gifts
# @param {Integer} k
# @return {Integer}
def pick_gifts(gifts, k)
    pq = gifts.sort.reverse
    
    while k > 0
        max_gifts = pq.shift
        if max_gifts == 1
            pq.push(1) # Add 1 back to the heap
            break
        end
        
        leave = Math.sqrt(max_gifts).floor
        pq.push(leave)
        pq.sort!.reverse!
        k -= 1
    end
    
    pq.sum
end

File: 2692-take-gifts-from-the-richest-pile/2692-take-gifts-from-the-richest-pile.js
/**
 * @param {number[]} gifts
 * @param {number} k
 * @return {number}
 */
var pickGifts = function(gifts, k) {
    const maxHeap = new MaxHeap(gifts);
    
    while (k > 0) {
        let maxGifts = maxHeap.pop();
        if (maxGifts === 1) 
        {
            maxHeap.push(1);
            break;
        }        
        let leave = Math.floor(Math.sqrt(maxGifts));
        maxHeap.push(leave);
        k--;
    }
    
    return maxHeap.sum();
};

class MaxHeap {
    constructor(arr) {
        this.heap = [...arr];
        this.heap.sort((a, b) => b - a);
    }
    
    pop() {
        return this.heap.shift();
    }
    
    push(val) {
        this.heap.push(val);
        this.heap.sort((a, b) => b - a);
    }
    
    sum() {
        return this.heap.reduce((a, b) => a + b, 0);
    }
}
